lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -589,35 +627,156 @@ def _validate_csr():
+                        return False
+                except cryptography.x509.ExtensionNotFound as e:
+                    return False
+            return True
+
+    def check(self, module, perms_required=True):
+        """Ensure the resource is in its desired state."""
+
+        state_and_perms = super(Certificate, self).check(module, perms_required)
 
         if not state_and_perms:
             return False
 
-        self.cert = crypto_utils.load_certificate(self.path)
+        self.cert = crypto_utils.load_certificate(self.path, backend=self.backend)
 
         if self.privatekey_path:
             try:
                 self.privatekey = crypto_utils.load_privatekey(
                     self.privatekey_path,
-                    self.privatekey_passphrase
+                    self.privatekey_passphrase,
+                    backend=self.backend
                 )
             except crypto_utils.OpenSSLBadPassphraseError as exc:
                 raise CertificateError(exc)
-            return _validate_privatekey()
+            return self._validate_privatekey()
 
         if self.csr_path:
-            self.csr = crypto_utils.load_certificate_request(self.csr_path)
-            if not _validate_csr():
+            self.csr = crypto_utils.load_certificate_request(self.csr_path, backend=self.backend)
+            if not self._validate_csr():
                 return False
 
         return True
 
 
+class SelfSignedCertificateCryptography(Certificate):
+    """Generate the self-signed certificate, using the cryptography backend"""
+    def __init__(self, module):
+        super(SelfSignedCertificateCryptography, self).__init__(module, 'cryptography')
+        self.notBefore = self.get_relative_time_option(module.params['selfsigned_not_before'], 'selfsigned_not_before')
+        self.notAfter = self.get_relative_time_option(module.params['selfsigned_not_after'], 'selfsigned_not_after')
+        self.digest = crypto_utils.select_message_digest(module.params['selfsigned_digest'])
+        self.version = module.params['selfsigned_version']
+        self.serial_number = x509.random_serial_number()
+        self.csr = crypto_utils.load_certificate_request(self.csr_path, backend=self.backend)
+        self._module = module
+
+        try:
+            self.privatekey = crypto_utils.load_privatekey(
+                self.privatekey_path, self.privatekey_passphrase, backend=self.backend
+            )
+        except crypto_utils.OpenSSLBadPassphraseError as exc:
+            module.fail_json(msg=to_native(exc))
+
+    def generate(self, module):
+        if not os.path.exists(self.privatekey_path):
+            raise CertificateError(
+                'The private key %s does not exist' % self.privatekey_path
+            )
+        if not os.path.exists(self.csr_path):
+            raise CertificateError(
+                'The certificate signing request file %s does not exist' % self.csr_path
+            )
+        if self.digest is None:
```

Comment:
I would move this `if` already into `__init__`, right after `self.digest` is computed.

Reply:
yep, agreed. Done